### PR TITLE
[python][torch] Add batch-first streaming datasets

### PR DESCRIPTION
**Pull request by XiaoHongbo-Hope**
### Purpose

Avoid per-row Python conversion by allowing PyPaimon Torch streaming datasets to yield PyArrow `RecordBatch` objects or dictionaries of Torch tensors.

- `batch_format="row"` remains the default; `"pyarrow"` and `"torch"` enable batch streaming.
- `batch_size=None` preserves reader batches; otherwise batches are combined or sliced.
- The default converter supports non-null numeric, boolean, and numeric fixed-size-list columns. Other types can use `to_tensor_fn`.
- Use `DataLoader(batch_size=None)` to avoid a second batching step.

Batch formats use one reader per DataLoader worker and require `prefetch_concurrency=1`, because Python producer threads cannot reliably cancel a reader blocked in I/O. A non-binding global limit preserves DataLoader worker parallelism when validated merged row counts, or physical row-count upper bounds, prove that all rows fit.

### Related implementations

[LanceDataset](https://github.com/lance-format/lance/blob/main/python/python/lance/torch/data.py) also streams Arrow batches and provides `batch_size`, `to_tensor_fn`, and native `batch_readahead`. PyIceberg provides [`to_arrow_batch_reader()`](https://py.iceberg.apache.org/reference/pyiceberg/table/#pyiceberg.table.DataScan.to_arrow_batch_reader), but not a PyTorch Dataset or Tensor conversion layer.

### Tests

- Torch tests: 25 passed
- Binding and non-binding multi-worker limits, merged-row limits, and Arrow offset-overflow regressions
- Flake8, Python 3.6 `py_compile`, and `git diff --check`

### CODE REVIEW

File: paimon-python/pypaimon/read/datasource/torch_dataset.py [hunk 1]
```diff
@@ -100,10 +104,35 @@ def _row_to_dict(self, offset_row) -> dict:
             row_dict[field_name] = value
         return row_dict
 
+    def _limit_covers_all_splits(self) -> bool:
+        limit = self.table_read.limit
+        if limit is None:
+            return True
+        total_rows = 0
+        for split in self.splits:
+            row_count = getattr(split, "row_count", None)
```
Comment (JingsongLi) [inline paimon-python/pypaimon/read/datasource/torch_dataset.py]: [P2] Use merged row counts before collapsing DataLoader workers

Split.row_count is the sum of physical file rows, so it can substantially exceed the logical output for deletion-vector and Data Evolution splits. For example, if two splits each have 10 physical rows but 4 merged rows and the limit is 8, the limit is non-binding; this code nevertheless totals 20 and routes every split to worker 0, unnecessarily regressing row, shuffle, and batch streaming from multi-worker to single-worker execution. TableRead._limit_covers_all_splits already uses split.merged_row_count() for the same proof. Please prefer a validated merged_row_count() when it is available, fall back conservatively when it is unknown, and add a non-binding test where physical and merged counts differ.

Comment (XiaoHongbo-Hope) [reply paimon-python/pypaimon/read/datasource/torch_dataset.py]: Fixed in cc695c0e0. The worker-limit proof now prefers merged_row_count() only when it is a valid nonnegative count no larger than the physical row count; otherwise it falls back to the physical upper bound. The regression covers 20 physical rows, 8 merged rows, and limit=8 retaining both workers.

**Review by JingsongLi — APPROVED** [review]
+1
